File: fchach/static/js/prodfield.js
function isInt(val) {
	if (isNumber(val) && (parseFloat(val) == parseInt(val))){
		return true;
	} 
	else {
		return false;
	}
};

function isNumber(val) {
	if(!val || (typeof val != "string" || val.constructor != String)) {
		return(false);
	}
	return !isNaN(new Number(val));
};

function check_line(){
	$prix = $("#prix").val().replace(',','.');
	$qte = $("#quantite").val()

		if (isNumber($prix) && isInt($qte)){
			$("#sstotal").val(String($prix*$qte).replace('.',','));
			return true;
		}
		else{
			reset();
			return false;
		}

};

function reset(){
	$("#prod_add").find("input").map(function(){ $(this).val(""); });
	$("#mat_opt[value='']")[0].selected=true;
	$("#nom_opt[value='']")[0].selected=true;
	$("#quantite").val("1");
	$("#sstotal").val("0");
}

function recal_total(){
	total=0;
	$(".sstotal").each(function(){
		total += parseFloat($(this).val().replace(',','.'));
	});
	$('#total').val(String(total).replace('.',','));
};

$(document).on("click", "#add_button", function(){
	selected_mat = $(".mat_opt").filter(":selected")[0].value
	selected_nom = $(".nom_opt").filter(":selected")[0].value	
	selected_mat_label = $(".mat_opt").filter(":selected")[0].label
	selected_nom_label = $(".nom_opt").filter(":selected")[0].label
	if (check_line() && $("#intitule").val() && selected_mat && selected_nom){
		prod = "prod"+i;

		$("#total_line").before($("#prod_add")
			.clone()
			.attr("id",prod)
			);

		// on change l'id des champs
		$("#"+prod).find("td > .intitule").attr("id", "intitule-"+i);	
		$("#"+prod).find("td > .intitule").attr("name", "intitule-"+i);

		$("#"+prod).find("td > .nomenclature").attr("id", "nomenclature-"+i);	
		$("#"+prod).find("td > .nomenclature").attr("name", "nomenclature-"+i);	
		$("#"+prod).find("td > .nomenclature").attr("rel", "tooltip");	
		$("#"+prod).find("td > .nomenclature").attr("title", selected_nom_label);	
		$("#"+prod).find("td > .nomenclature").after("<input type='hidden' name='nomenclature-"+i+"' value='"+selected_nom+"' />");	
		$("#"+prod).find("td > .nomenclature").find("option[value="+selected_nom+"]")[0].selected = true;
		$("#"+prod).find("td > .nomenclature").find("option").map(function(){ $(this).attr("id", "nom_opt-"+i); });

		$("#"+prod).find("td > .matiere").attr("id", "matiere-"+i);	
		$("#"+prod).find("td > .matiere").attr("name", "matiere-"+i);	
		$("#"+prod).find("td > .matiere").attr("rel", "tooltip");	
		$("#"+prod).find("td > .matiere").attr("title", selected_mat_label);	
		$("#"+prod).find("td > .matiere").after("<input type='hidden' name='matiere-"+i+"' value='"+selected_mat+"' />");	
		$("#"+prod).find("td > .matiere").find("option[value="+selected_mat+"]")[0].selected = true;
		$("#"+prod).find("td > .matiere").find("option").map(function(){ $(this).attr("id", "mat_opt-"+i); });

		$("#"+prod).find("td > .prix").attr("id", "prix-"+i);	
		$("#"+prod).find("td > .prix").attr("name", "prix-"+i);	
		$("#"+prod).find("td > .qte").attr("id", "quantite-"+i);	
		$("#"+prod).find("td > .qte").attr("name", "quantite-"+i);	
		
		// readonly
		$("#"+prod).find("input").map(function(){ $(this).attr("readonly", "readonly"); });
		$("#"+prod).find("select").map(function(){ $(this).attr("disabled", "disabled"); });

		$("#"+prod).find("#add_button").children().remove();
		$("#"+prod).find("#add_button").append("<i class=icon-minus></i>");	
		$("#"+prod).find("#add_button").attr("class", "suppr");	
		$("#"+prod).find("#add_button").attr("id", "suppr-"+i);	

		reset();
		recal_total();

		i++;
	}
	else{
		$("#produits").before('<div class="alert alert-error"><a class="close" data-dismiss="alert" href="#">×</a><h4 class="alert-heading">Erreur</h4>Remplissez les champs correctement !</div>');
	}
});

$(document).on("click", ".suppr", function(){ $(this).parent().parent().remove(); recal_total(); });

$(document).on("change", ".prix,.qte", check_line);


/* Reponsive table */

$(document).ready(function() {
	  var switched = false;
	  var updateTables = function() {
	    if (($(window).width() < 980) && !switched ){
	      switched = true;
	      $("table.responsive").each(function(i, element) {
		minTable($(element));
	      });
	      return true;
	    }
	    else if (switched && ($(window).width() > 980)) {
	      switched = false;
	      $("table.responsive").each(function(i, element) {
		maxTable($(element));
	      });
	    }
	  };
	   
	  $(window).load(updateTables);
	  $(window).bind("resize", updateTables);
	   
		
	function minTable(table){
		table.find(".input-small").removeClass("input-small").addClass("input-mini butsmall");
		table.find(".input-normal").removeClass("input-normal").addClass("input-small butnormal");
		table.find("th#mat_xlab").text("XLAB");
		table.find("th#nomenclature").text("Nom.");
		}
		
	function maxTable(table) {
		table.find(".butnormal").removeClass("input-small butnormal").addClass("input-normal");
		table.find(".butsmall").removeClass("input-mini butsmall").addClass("input-small");
		table.find("th#mat_xlab").text("Matière XLAB");
		table.find("th#nomenclature").text("Nomenclature");
		}

});

File: fchach/prodfield.py
# -*- coding: utf-8 -*-
"""
A newforms widget and field to allow multiple file uploads.

Created by Edward Dale (www.scompt.com)
Modified by Rémi Cailletaud
Released into the Public Domain
"""

from django.forms.fields import Field
from django.db import models
from django.forms.widgets import Widget
from django.forms.util import ValidationError
from compta.models import Nomenclature, NomenclatureCat, MatiereXlab, MatiereXlabCat

def clean_data(data):
    data_cleaned = {}
    for k in data.items():
        if k[0].partition("-")[2] not in data_cleaned:
            data_cleaned[k[0].partition("-")[2]] = dict()
        data_cleaned[k[0].partition("-")[2]][k[0].partition("-")[0]] = k[1][0]
    return data_cleaned


class ProdWidget(Widget):
    """
    Widget pour la liste de produits
    """

    def __init__(self, attrs=None):
        self.nomenclatures=Nomenclature.objects.all().order_by('nom')
        super(ProdWidget, self).__init__(attrs)


    def render_mat_opt(self, selected=None):
        opt_str = "<option id='mat_opt' class='mat_opt' value=''>---------</option>"
        xlab_categories = MatiereXlabCat.objects.all().order_by('numero')
        for xlab_cat in xlab_categories:
            opt_str += "<optgroup label='%(categorie)s'>" % { 'categorie': xlab_cat }
            matieres = MatiereXlab.objects.filter(categorie=xlab_cat).order_by('numero')
            for matiere in matieres:
                if selected and selected == matiere.id:
                    opt_str += u"<option id='mat_opt' class='mat_opt' value='%(id)s' selected=selected>%(matiere)s<option>" \
                                    % { 'id': matiere.id, 'matiere': matiere }
                else:
                    opt_str += u"<option id='mat_opt' class='mat_opt' value='%(id)s'>%(matiere)s</option>" \
                                    % { 'id': matiere.id, 'matiere': matiere }
        return opt_str

    def render_nom_opt(self, selected=None):
        opt_str = "<option id='nom_opt' class='nom_opt' value=''>---------</option>"
        nom_categories = NomenclatureCat.objects.all().order_by('numero')
        for nom_cat in nom_categories:
            opt_str += "<optgroup label='%(categorie)s'>" % { 'categorie': nom_cat }
            nomenclatures = Nomenclature.objects.filter(categorie=nom_cat).order_by('numero')
            for nomenclature in nomenclatures:
                if selected and selected == nomenclature.id:
                    opt_str += u"<option id='nom_opt' class='nom_opt' value='%(id)s' selected=selected>%(nomenclature)s<option>" \
                                    % { 'id': nomenclature.id, 'nomenclature': nomenclature }
                else:
                    opt_str += u"<option id='nom_opt' class='nom_opt' value='%(id)s'>%(nomenclature)s</option>" \
                                    % { 'id': nomenclature.id, 'nomenclature': nomenclature }
        return opt_str


    def render(self, name, value, attrs=None):
        """
        Renders the ProdWidget
        Should not be overridden.  Instead, subclasses should override the
        js, link, and/or fields methods which provide content to this method.
        """
        final_attrs = self.build_attrs(attrs, type='text', name=name+'[]')
        if 'disabled' not in final_attrs:
            final_attrs['disabled'] = False
        js = self.js(name, value, final_attrs)
        fields = self.fields(name, value, final_attrs)

        return js+fields


    def add_row(self, num, intitule, matiere, nomenclature, prix, quantite, disabled):
        """
        Recharge une ligne de la table produits
        """
        row = u"""<tr id=prod%(num)s>
                    <td><input id="intitule-%(num)s" class="intitule" type="text" name="intitule-%(num)s" readonly="readonly" value="%(produit)s"></td>
                    <td><select class="matiere input-small" id="matiere-%(num)s" disabled="disabled">%(mat_options)s</select>
                        <input type="hidden" name="matiere-%(num)s" value="%(mat_id)s" /></td>
                    <td><select class="nomenclature input-small" id="nomenclature-%(num)s" disabled="disabled">%(nom_options)s</select>
                        <input type="hidden" name="nomenclature-%(num)s" value="%(nom_id)s" /></td>
                    <td><input id="prix-%(num)s" class="prix input-mini" type="text" name="prix-%(num)s" readonly="readonly"  value="%(prixht)s"></td>
                    <td><input id="quantite-%(num)s" class="qte input-mini" type="text" name="quantite-%(num)s" readonly="readonly"  value="%(quantite)s"></td>
                    <td><input id="sstotal" class="sstotal input-mini" type="text" readonly="readonly" value="%(sstotal)s">
                    </td>""" % {'num': num, 'produit': intitule,
                                'mat_options': self.render_mat_opt(matiere), 'mat_id': matiere,
                                'nom_options': self.render_nom_opt(nomenclature), 'nom_id': nomenclature,
                                'prixht': str(prix).replace('.',','),
                                'quantite': quantite, 'sstotal': str(prix*quantite).replace('.',',') }

        if not disabled :
            row += """<td>
                    <a id="suppr-%(num)s" class="suppr"><i class=icon-minus></i></a>
                    </td>"""

        return row + """</tr>"""

    def fields(self, name, value, attrs=None):
        """
        Renders the necessary number of input boxes.
        """
        table = u"""
                        <table  class="table-bordered table-striped table-condensed responsive span8" id="produits">
                        <thead>
                        <tr><th>Intitulé</th><th id="mat_xlab" >Matière XLAB</th><th id="nomenclature">Nomenclature</th><th>Prix</th><th>Quantit&eacute;</th><th>Sous-total</th></tr>
                        </thead>
                        <tbody>
                        """

        if not attrs['disabled']:
            table += u"""<tr id="prod_add">
                         <td><input class="intitule input-normal" id="intitule"  type="text"/></td>
                         <td><select class="matiere input-small" id="matiere">%(mat_options)s</select></td>
                         <td><select class="nomenclature input-small" id="nomenclature">%(nom_options)s</select></td>
                         <td><input class="prix input-mini" id="prix"  type="text"/></td>
                         <td><input class="qte input-mini" id="quantite" value="1" type="text"/></td>
                         <td><input class="sstotal input-mini" id="sstotal" readonly="readonly" value="0"  type="text"/></td>
                         <td><a id="add_button"><i class="icon-plus"></i></a></td></tr>
                         <tr><td colspan="5">Cliquez sur <i class="icon-plus"></i> pour ajouter les produits au panier </td>
                         </tr> """ % {'mat_options': self.render_mat_opt(), 'nom_options': self.render_nom_opt()}

        if isinstance(value,list):
            for val in value:
               table += self.add_row(value.index(val)+1, val.produit, val.matiere_xlab.id, val.nomenclature.id, val.prixht, val.quantite, attrs['disabled'])
        if isinstance(value,dict):
            # si recharge après erreur de formulaire
            #import ipdb; ipdb.set_trace()
            for val in clean_data(value).values():
               table += self.add_row(clean_data(value).values().index(val)+1, val['intitule'], int(val['matiere']), int(val['nomenclature']), \
                                        float(val['prix'].replace(',','.')), int(val['quantite']), attrs['disabled'])
        table += '''
                    <tr id="total_line"><td colspan="5"><strong>Total:</strong></td><td><input id="total" class="input-mini" readonly="readonly"/></td></tr>
                     </tbody></table>
                '''
        return table

    def js(self, name, value, attrs=None):
        """
        Javascript qui va bien.
        """
        if attrs['disabled']:
            return ''
        else:
            return u'''
        <script type="text/javascript" src="/static/js/prodfield.js"></script>
        <script type="text/javascript">window.onload=function(){ i=%(num)s; recal_total(); }</script>
        ''' % {'num': len(value)+1 }

    def value_from_datadict(self, data, files, name):
        """
        On récupère les valeurs pour les champs produits
        """
        my_data = {}
        for d in data.lists():
            if d[0].startswith('prix') or d[0].startswith('quantite') or d[0].startswith('intitule') or \
                    d[0].startswith('matiere') or  d[0].startswith('nomenclature'):
                my_data[d[0]] = d[1]
        return my_data

class ProdField(Field):
    """
    Field pour les produits
    """
    __metaclass__ = models.SubfieldBase

    def __init__(self, *args, **kwargs):
        self.widget = kwargs['widget']
        super(ProdField, self).__init__(*args, **kwargs)

    def clean(self, data):
        """
        Vérifie qu'on a au moins un produit, lève une exception sinon
        Trie les produits par leur "id" dans la fiche
        """
        super(ProdField, self).clean(data)
        if not data:
            raise ValidationError(u"Vous n'avez précisé aucun produit !")
        else:
            return clean_data(data)
